Simplify index expressions constructed in loop flattening - #51173

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52882 Simplify index expressions constructed in loop flattening - #51173**

Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

huiguoo committed:
Simplify index expressions constructed in loop flattening - #51173

[ghstack-poisoned]

bertmaher reviewed (commented):
This is cool!

ZolotukhinM commented:
Thanks for working on this, this would be extremely useful as we start exercising loop transformations more!

I wonder if we could make it more robust - the current approach seems a bit fragile and could probably not detect some cases, which are similar to, but not exactly match the pattern.

I was thinking that maybe we could break this big pattern into a series of smaller more generic patterns - that would make the entire solution more stable. Below is an illustration of what I mean.

As you commented on the issue, we're dealing with an expression in this form:
```
(t/(K1 * K2 * K3) %K4) * (K1 * K2 * K3) +
(t/(K1 * K2) %K3) * (K1 * K2) +
(t/K1 %K2) * K1 +
t%K1
```
Note, that it is equivalent to
```
(t/(K1 * K2 * K3) %K4) * (K1 * K2 * K3) +
(t/(K1 * K2) %K3) * (K1 * K2) +
t%(K1*K2)
```
And if we now substitute "K1*K2" with "K_12", then we'll get
```
(t/(K_12 * K3) %K4) * (K_12 * K3) +
(t/K_12 %K3) * K_12 +
t%K_12
```
Now we can substitute `(t/K_12 %K3) * K_12 + t%K_12` with `t%(K_12*K3)`, which would lead to
```
(t/(K_12 * K3) %K4) * (K_12 * K3) +
t %(K_12 * K3)
```
Now we can replace `K_12*K3` with `K_123` and keep repeating this process until we simplified the entire expression - for the one from the example we should end up with just `t`.

How can we get there with small steps?
We can consider performing two transformations on an expression:
1) `(A/B %C) * B` ---> `(A/(B*B)) % (C*B)` ---> `RoundOff(A, B) % (C*B)`
2) `RoundOff(A,B) % (C*B) + A%B` ---> `A % (C*B)`

If we apply these rules iteratively to the original expression, I believe we will fold it to the end (provided that we properly detect multipliers where they are needed). And for more complicated cases this could at least partially simplify them, even if the whole pattern doesn't match.

Curious to hear what are your thoughts on this?

huiguoo commented:
@ZolotukhinM I like this recursive algorithm! I totally agree that we need to be more general when folding flattened expressions and this recursive solution seems the right way to do it. 

In the transformations, we can avoid distributing the multiplier B (i.e., `(A/B % C) * B` -> `(A/B*B) % (C*B)`) by extending the definition of class `RoundOff` from 
`RoundOff(A, B) = A/B*B` to 
`RoundOffOverC(A, B, C) = (A/B%C) * B` (there could be a better name than `RoundOffOverC`).

With the extended `RoundOff`, we can do the reduction
`(A/B%C) * B + A%B = A%(B*C)` 
using the rule
`RoundOffOverC(A, B, C) + A%B = A%(B*C)`.

The reduction on the index expression of 2-d flattened buf
`A/B * B + A%B`
can be reduced in the following process,
`RoundOffOverC(A, B, Inf) + A%B = A%(B*Inf) = A`.

Does it seem more neat?

ZolotukhinM commented:
Awesome, glad that you like the idea :)

Regarding the `RoundOffOverC`, do you expect that it'd help with a simpler or more robust implementation? My two main concerns with using it are:
1) The semantics of it is more complicated than `RoundOff` and I'm trying to understand how hard would it be to grasp it later (not for us, but for people who would read the code without the context we have now). If we go that route, we need to make sure it's really really well-documented and clear what exactly such nodes do.
2) I'm a bit worried that representing `Inf`s might be hard (e.g. if we reserve a special value like `0` to represent it, then the semantics of the op will be even more complicated. If we introduce a new expr type, then we'll have to extend number of our core abstractions just for this optimization, which we ideally should try to avoid).

huiguoo commented:
I agree that the semantics of `RoundOffOver` seems complicated and could be hard to grasp for other people. The main reason I bring it up is that using it we can formulate `(A/B %C) * B` into one data structure. This can avoid transforming `(A/B % C) * B` to `RoundOff(A, B) % (B*C)` which may introduce complexities for other simplifications that consider B as a common factor. 

For example,  `(A/B % C) * B + (A/B/C) * C * B` can be simplified in the following transformations,
`= (A/B % C + (A/B/C) * C) * B`
`= (A/B % C, RoundOff(A/B, C)) * B`
`= (A/B) * B`

If we transform `(A/B % C) * B` to `RoundOff(A, B) % (B*C)` before this simplification, we need to transform it back. But this example may be unusual. In that case, we don't need to be bothered adding another expr or redefining `RoundOff`. I'll go ahead and implement the recursive algorithm you described above.

ZolotukhinM commented:
Yeah, that's an interesting example! And I do see how your proposal is more general and allows us to handle this case as well - let's just weigh in practical pros and cons for now and stick with a simpler solution until we see a real case needing a more general one.

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
+  // any further.
+  while (!mods.empty() && repeat) {
+    repeat = false;
+    for (std::vector<const Term*>::iterator mit = mods.end() - 1;
+         mit != mods.begin() - 1;
+         mit--) {
```
Why do we iterate to `mods.begin() - 1` here?

huiguoo replied:
Why in reversed order? In one iteration, if the mod term can be merged with other terms, I'll immediately do merging and erase it from the vector without breaking the for loop. So iterate mod items from `mods.end()-1` to `mods.begin()`; the exit check would be `mit != mods.begin()-1` (i.e., `mit >= mods.begin()`).

ZolotukhinM replied:
I understand why it is in reversed order, my question was why `mods.begin() - 1` instead of `mods.begin()` - I was confused by '-1', but now I got it, thanks!

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
+      const Expr* mod_lhs = IRSimplifier::simplify(mod->lhs());
+      const Expr* mod_rhs = IRSimplifier::simplify(mod->rhs());
+      bool merged = false;
+      for (std::vector<const Term*>::iterator mrit = mod_rounds.end() - 1;
+           mrit != mod_rounds.begin() - 1;
+           mrit--) {
```
Same question.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
   }
 
-  // Keep remaining Mods.
+  // Keep remaining Mods, ModRounds and RoundOffs.
   for (auto* m : mods) {
     others.push_back(m);
   }
+  for (auto* mr : mod_rounds) {
+    others.push_back(mr);
+  }
+  for (auto* r : rounds) {
+    others.push_back(r);
+  }
```
Nit: probably we could use `mods.insert` instead of these loops.

ZolotukhinM reviewed (commented):
This looks very good and this is an awesome optimization! I left some remarks below, once they are addressed I think the PR would be good to go! 👍

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
-// Searches the polynomial for Terms that can be merged in the Round + Mod
-// pattern: (x/y) * y + x % y => RoundOff(x,y) + Mod(x, y) => x.
+// A ModRound is a div-mod-mul in which the divisor in div and multiplier in mul
+// are identical. Return (scalar, denominator, divisor, mod_divisor). Example:
+// ((t/7)%9)*7 -> (1, t, 7, 9)
```
For `(a/b%c)*d`, if d%b != 0 then it's not a ModRound pattern; otherwise the scalar=`d/b`, denominator=`a`, divisor=`b` and mod_divisor=`c`.

ZolotukhinM replied:
Could we reflect this to the comment more explicitly?

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
+
+  for (auto* m : e->variables()) {
+    if (m->expr_type() == IRNodeType::kMod) {
+      // TODO: currently only identify terms with one variable being mod; it is
+      // possible to extend this if we have to handle terms like (t/(x%2 * y) %
+      // z) * (x%2 *y).
```
It's not a blocker from simplifying Nd expressions. If such terms do not exist before simplification, we would not generate such terms in the simplification iterations. Each iteration generates either (1) a mod term (`x%(y*z)`) by merging a modround term (`(x/y%z)*y`) and a mod term (`x%y`), or (2) a `x` term by adding `Roundoff(x, y)` and `x%y` - no new modround terms are generated. If `y` is not originally a mod we should be fine.

huiguoo commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,81 +1717,244 @@ const Expr* polyGCD(const Polynomial* poly) {
+  // Deny cases in which divisor!=multiplier.
+  HashProvider& hasher = e->hasher();
+  if (hasher.hash(divisor) != hasher.hash(multiplier)) {
+    if (divisor->isConstant() && multiplier->isConstant()) {
```
Yeah, I think it would be great if we can extend it. 
As we discussed, there are currently no simplifications like `x*2/x -> 2` (because x could be 0) so I'm not able to extract `2` out of the `multiplier` to be `scalar`. But we can extend division to 2 versions: 1) division for customers that has to be strictly simplified  and 2) division we introduced in our transformations  which can be simplified without considering 0s, e.g., `Div_nonzero`. The second division will be only used to facilitate our transformations.

ZolotukhinM replied:
Yeah, that sounds good! Please put a TODO in the code with these thoughts and we can return to that later.

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -2691,15 +2691,302 @@ TEST(Simplify, SimplifyRoundModPatternMultivar) {
+  }
+
+  {
+    // t/2 /7 % 9 * 7 + t/2 % 7 => t/2 % 63
```
Would this work if we had `t/14 % 9 * 7 + t/2 % 7 => t/2 % 63`?

huiguoo replied:
Good point! Added support for such cases.

ZolotukhinM commented on test/cpp/tensorexpr/test_simplify.cpp:
```diff
@@ -2691,15 +2691,302 @@ TEST(Simplify, SimplifyRoundModPatternMultivar) {
+    IS_VAR_WITH_NAME(simplified.node(), "io_flat");
+  }
+
+  { // 5D: i0_flat / (k * l * n * m)  * (m * n * l * k) +
+    // (i0_flat / (l * n * m) % k)  * m * n * l +
+    // (i0_flat / (n * m) % l) * m * n +
+    // (i0_flat / m % n)  * m +
+    // i0_flat % m => io_flat
```
This is awesome! 😍

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -1717,80 +1717,252 @@ const Expr* polyGCD(const Polynomial* poly) {
+      const Expr* denom,
+      const Expr* divisor,
+      const Expr* mod_divisor)
+      : scalar(scalar),
```
That clang-tidy complaint seems easy to fix.

ZolotukhinM reviewed (approved):
Great job!
![image](https://user-images.githubusercontent.com/4588111/110226181-d8f98e00-7ea1-11eb-9c44-3fc7fc695c72.png)

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

ezyang commented:
This broke ASAN. Unlanding.

huiguoo commented:
@ezyang Can you share the cmd that breaks ASAN and the error you see?

bertmaher commented:
Here's a dashboard where you can find the broken build if you squint: https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master

huiguoo committed:
Update on "Simplify index expressions constructed in loop flattening - #51173"


Differential Revision: [D26676150](https://our.internmc.facebook.com/intern/diff/D26676150)

[ghstack-poisoned]

ezyang commented:
ASAN failed on PR CI too, which you can see if you click on the icon ont he last commit before landing. https://app.circleci.com/pipelines/github/pytorch/pytorch/283747/workflows/03b5e788-38aa-40cd-9fe6-436a5474a1ab/jobs/11455054

huiguoo commented:
Thank you @ezyang ! I'll fix the errors :)
